### PR TITLE
Initial Jammy Tiny Implementation

### LINKED ISSUE #1: Implement Paketo Stacks RFC 0004: Jammy Jellyfish
[Paketo Stacks RFC 0004](https://github.com/paketo-buildpacks/rfcs/blob/ba9273ded6c65274276677b1ad5db5e61029060d/text/stacks/0004-jammy-jellyfish.md) states that Paketo should support a "tiny" stack on top of Ubuntu Jammy Jellyfish. To implement this RFC:

- [ ] Determine which OS packages should be installed on the Tiny Jammy stack. These may be identical to those installed on the bionic tiny stack. (See [the bionic tiny repo](https://github.com/paketo-buildpacks/tiny-stack-release))
- [ ] Create a [stack descriptor file](https://github.com/paketo-buildpacks/rfcs/blob/ba9273ded6c65274276677b1ad5db5e61029060d/text/stacks/0003-stack-descriptor.md) for this stack and check it into this repo. The descriptor should have the following attributes:
    - stack id: `io.buildpacks.stacks.jammy.tiny`
    - the UID for `build` and `run` phases should differ (but be in the same group)
- [ ] Set up automation on this repo that builds the stack using `jam create-stack` and publishes the stack to the appropriate tags:
    - `paketobuildpacks/build-jammy-tiny:latest`
    - `paketobuildpacks/build-jammy-tiny:1.2.3`
    - `paketobuildpacks/run-jammy-tiny:latest`
    - `paketobuildpacks/run-jammy-tiny:1.2.3`
- [ ] Create automation that allows the repository to receive USN updates
- [ ] Create automation that generates release notes with similar information to the Bionic tiny stack release notes (e.g. [this release](https://github.com/paketo-buildpacks/tiny-stack-release/releases/tag/1.3.44))

### PR DESCRIPTION
**Pull request by fg-j**
## Summary

Resolves #1 

To Do:

- [x]  Create an action that searches previous releases for patched USN JSON files. Pass this JSON as input to the Get USNs action.
- [x] Create an action that takes two package receipts as input and produces a package diff split into lists of added and removed packages as output. May be very similar to [this existing action](https://github.com/paketo-buildpacks/stacks/tree/main/actions/get-package-diff).
- [x] Add a check to see if packages have been REMOVED since the previous release. Fail the release workflow if so.
- [x] Update the create release workflow to only cut a release if
    - [x] the acceptance tests pass AND
    - [x] either of the build/run images are different than the previous release OR
    - [x] the force-creation flag is set
- [x] Fix the Increment Tag action to allow re-tagging a commit that's already been tagged (https://github.com/paketo-buildpacks/github-config/pull/474) 
- [x] Do not report the build as failed if there are no changes - just do not create a release.
- [x] Actually run the integration tests
- [x] TODOs in stack/stack.toml
- [x] TODOs in metadata_test.go
- [x] TODOs in actions/find-release-asset/entrypoint
- [x] TODOs in actions/find-and-download-asset/action.yml
- [x] TODOs in scripts/receipts.sh
- [x] TODOs in actions/get-usns/entrypoint/main.go
- [x] TODOs in .github/workflows/create-release.yml
- [x] Remove extraneous comments / dead code

---

Outstanding issues that can/should be tackled later
- [ ] Evaluate rewriting actions in golang (with tests as appropriate) https://github.com/paketo-buildpacks/jammy-tiny-stack/issues/4
- [ ] Add test coverage to the `get-usns` action https://github.com/paketo-buildpacks/jammy-tiny-stack/issues/3
- [ ] Refactor the workflows and actions into a common location so we can re-use them in other stacks
- [x] Evaluate how re-usable the notify on failure step is - can we abstract that action and re-use it in other pipelines? https://github.com/paketo-buildpacks/jammy-tiny-stack/issues/5
- [ ] Evaluate whether we should validate the uploaded release artifacts after we upload them to the newly-created release https://github.com/paketo-buildpacks/jammy-tiny-stack/issues/6

## Checklist

* [x] I have viewed, signed, and submitted the Contributor License Agreement.
* [x] I have linked issue(s) that this PR should close using keywords or the Github UI (See [docs](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue))
* [x] I have added an integration test, if necessary.
* [x] I have reviewed the [styleguide](https://github.com/paketo-buildpacks/community/blob/main/STYLEGUIDE.md) for guidance on my code quality.
* [x] I'm happy with the commit history on this PR (I have rebased/squashed as needed).

### CODE REVIEW

File: .github/labels.yml
Comment (loewenstein) [inline .github/labels.yml:1]: Are those labels used across the org? Might be worth checking if labels are supported in an org-wide `.github` repo.

Comment (sophiewigmore) [reply .github/labels.yml:1]: We do this for builders (and similarly for buildpacks): https://github.com/paketo-buildpacks/github-config/blob/main/builder/.github/labels.yml
We could do the same for stacks.

**Review by sophiewigmore — DISMISSED** [review]

**Comment by sophiewigmore** [comment]
approval pending commented out stuff is put back in  or removed :)

**Comment by fg-j** [comment]
Needs https://github.com/paketo-buildpacks/github-config/pull/474 to be merged.

**Comment by robdimsdale** [comment]
The only outstanding TODOs are switching the repository for the actions from `fg-j` to `paketo-buildpacks`, but we once we do this the workflow on this repo will fail until the PR is merged. So it should be the very last thing that we do to allow us to continue getting feedback on this workflow as long as possible.

File: .github/labels.yml
Comment (fg-j) [reply .github/labels.yml:1]: Seems like it's possible to set labels org-wide: https://docs.github.com/en/organizations/managing-organization-settings/managing-default-labels-for-repositories-in-your-organization

Seems like this would have to be done by someone with org admin permissions.

**Comment by fg-j** [comment]
I took some time testing out different scenarios with the create-release workflow. Links to representative workflows added. First scenario hasn't happened yet with the completed pipeline.

| Scenario                              |  **Expected Outcome**              | Link                                                             |
|---------------------------------------|---------------------------|------------------------------------------------------------------|
| cron runs; there's a new USN to patch |  Release                  |                                                                  |
| cron runs; no new USNs                |  Halt after USNs job      | https://github.com/fg-j/jammy-tiny-stack/actions/runs/2470338131 |
| stack definition has changed          |  Release                  | https://github.com/fg-j/jammy-tiny-stack/actions/runs/2470616244 |
| trivial commit                        |  Skip Release             | https://github.com/fg-j/jammy-tiny-stack/actions/runs/2470814438 |
| trivial commit; forced release        |  Release                  | https://github.com/fg-j/jammy-tiny-stack/actions/runs/2470857613 |
| package is removed                    |  Fail release; file issue | https://github.com/fg-j/jammy-tiny-stack/actions/runs/2470403990 |
| package is removed; forced release    |  Fail release; file issue | https://github.com/fg-j/jammy-tiny-stack/actions/runs/2470448617 |
| change fails tests                    |  Fail release; file issue | https://github.com/fg-j/jammy-tiny-stack/actions/runs/2470534528 |

File: actions/generate-package-list/action.yml
Comment (fg-j) [inline actions/generate-package-list/action.yml:2]:
```suggestion
description: 'Generates a list of packages (without versions) given a dpkg -l style package receipt'
```

File: go.mod
Comment (fg-j) [inline go.mod:10]: Why are we depending on 2 versions of packit?

Comment (fg-j) [reply go.mod]: Fixed.

**Comment by robdimsdale** [comment]
@paketo-buildpacks/stacks-maintainers @ryanmoran we would love to get your eyes on this when you get chance.

File: actions/get-usns/entrypoint/main.go
Comment (ryanmoran) [inline actions/get-usns/entrypoint/main.go:62]: I wonder if it would be possible to get rid of the `patched-usns.json` file if we input the packages and their versions to this action. In that case, you could grab the USNs from some time period, maybe the last month, and then just compare affected packages and their versions. If you haven't upgraded to at least the version specified in the USN, then consider it an unpatched USN.

**Review by ryanmoran — COMMENTED** [review]
@fg-j & @robdimsdale, the broad strokes on this PR look great! I have some concerns about the size and complexity of the `create-release.yml` workflow though. I've made some comments in a couple of areas that I hope could reduce some of that complexity. Let me know what you think.

File: scripts/receipts.sh
Comment (fg-j) [reply scripts/receipts.sh:133]: This sounds great! Standardized across distros and machine-readable would be an improvement, IMO.

File: .github/workflows/create-release.yml
Comment (fg-j) [reply .github/workflows/create-release.yml:5]: It sure is! Fortunately, in the common case where there are no USNs, only the `Poll` job runs. This is still a  concern as stacks scale, though.

FWIW this cron schedule is  less frequent than [the current USN automation](https://github.com/paketo-buildpacks/stack-usns/blob/9b9eea66971ecb791b5c7d9d2b8c6f8db9267ba6/.github/workflows/get-usns.yml#L5). 

I worry about running this job too _infrequently_ since our current source of USNs truth only holds the 10 most recent USNs.

File: scripts/receipts.sh
Comment (robdimsdale) [reply scripts/receipts.sh:133]: I agree that an output in a generic SBOM format would be an improvement over `dpkg -l` format (assuming that consumers of this are OK with this breaking change in the release notes. I think that should be fine).

While I would certainly like to see this change, I think I'd prefer not to block this PR on `jam create-stack` gaining this feature. We can open it as an issue on github.com/paketo-buildpacks/jam and once the feature is implemented we can improve these scripts.

Sound good?

Comment (ryanmoran) [reply scripts/receipts.sh:133]: Sounds good to me.

File: .github/workflows/create-release.yml
Comment (robdimsdale) [reply .github/workflows/create-release.yml:5]: Yeah, we actually decreased the frequency from every 5 minutes to every 10 minutes. We might be able to get away with polling less frequently if we can be confident that the USN feed we receive still contains the last seen USNs. See the comment below about USN feed architecture.

If we stick with this architecture, the polling frequency should be about equal to number of USNs per feed (i.e. 10) / maximum rate of USN generation. Obviously the rate of USN generation isn't fixed, so this will always be a statistical approach.

Comment (fg-j) [reply .github/workflows/create-release.yml:5]: For now, let's stick with the cron schedule? We can come back and adjust this if it causes problems.

File: actions/get-usns/entrypoint/main.go
Comment (fg-j) [reply actions/get-usns/entrypoint/main.go:62]: Filed https://github.com/paketo-buildpacks/jammy-tiny-stack/issues/7 to investigate this.

**Review by ryanmoran — APPROVED** [review]

**Comment by fg-j** [comment]
❗ quick pause!

I'd like to propose we switch to using `syft` to generate package receipts/SBOMs before this merges!

File: README.md
Comment (fg-j) [reply README.md:16]: Good eye!

**Comment by sophiewigmore** [comment]
Tested locally on my Mac workstation and found logging/documentation helpful:
- `scripts/create.sh`
- `scripts/receipts.sh`
- `scripts/test.sh`
Also found the create-release workflow steps to be pretty clear. I think I would benefit as a maintainer having the different release criteria outlined somewhere (i.e. when an auto-release can happen vs when pressing the big red button is needed), however I don't think that needs to happen here. I left one nitpick change and then I'm personally good to approve this in and see how it goes maintaining it and using the workflows in reality.

**Comment by sophiewigmore** [comment]
@robdimsdale @fg-j it also looks like the Poll USNs job is failing: https://github.com/fg-j/jammy-tiny-stack/runs/6934783586?check_suite_focus=true

**Comment by sophiewigmore** [comment]
I just merged in https://github.com/paketo-buildpacks/github-config/pull/477 also, if you wish to update the workflow to use the issue-filing action. Up to you though

**Comment by fg-j** [comment]
@sophiewigmore Let's just get this PR in and we can tidy up the failure alert later.

**Review by sophiewigmore — APPROVED** [review]

**Comment by sophiewigmore** [comment]
@ryanmoran anything else you'd like on this? Otherwise I'm merging it in

**Comment by ryanmoran** [comment]
@sophiewigmore, nothing that should block this going in. Merge at will.

**Review by ryanmoran — APPROVED** [review]

**Pull request merged**
